### PR TITLE
Switch to using QP for n(z) information throughout

### PR DESCRIPTION
**Pull request by joezuntz**
This makes extensive changes to use QP objects for n(z) information throughout the code.

This PR includes #317 so should only be considered after that.

it also depends on QP PR [183](https://github.com/LSSTDESC/qp/pull/183).

### CODE REVIEW

File: txpipe/rail/summarize.py [hunk 1]
```diff
@@ -121,105 +122,66 @@ def run(self):
...
-            if self.config["zmax"] > 0:
-                ax.set_xlim(0, self.config["zmax"])
-            ax.set_ylim(0, None)
-            ax.set_xlabel("z")
-            ax.set_ylabel("n(z)")
+
+        zmax = self.config["zmax"]
+        nz = self.config["nz"]
+        z = np.linspace(0, zmax, nz)
+        print(z)
```
Comment (chihway) [inline txpipe/rail/summarize.py:165]: remove this? or add info on what is being printed

File: txpipe/rail/summarize.py [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
     name = "PZRailSummarize"
     inputs = [
-        ("tomography_catalog", HDFFile),
+        ("tomography_catalog", TomographyCatalog),
         ("photometry_catalog", HDFFile),
         ("model", PickleFile),
 
     ]
     outputs = [
         # TODO: Change to using QP files throughout
```
Comment (chihway) [inline txpipe/rail/summarize.py:15]: remove?

File: txpipe/photoz_stack.py [hunk 3]
```diff
@@ -1,496 +1,282 @@
...
 
-        # Save again but for the number of bins in the tomography catalog
-        with self.open_input("shear_tomography_catalog") as tomo_file:
-            nbin_source = tomo_file["tomography"].attrs["nbin"]
 
-        return z, nbin_source
 
 
-class TXPhotozLensStack(TXPhotozSourceStack):
+class TXTruePhotozStack(PipelineStage):
```
Comment (chihway) [inline txpipe/photoz_stack.py:128]: is there a version where we are using e.g. a z_mc value? that would be essentially the same as this class but histograming a different column (though I guess qp does not have that option now, and maybe not worth having...?)

File: examples/lensfit/pipeline.yml [hunk 4]
```diff
@@ -95,6 +100,7 @@ inputs:
     # This file comes with the code
     fiducial_cosmology: data/fiducial_cosmology.yml
     flow: data/example/inputs/example_flow.pkl
+    none: none
```
Comment (chihway) [inline examples/lensfit/pipeline.yml:103]: just curious, what is this?!

File: txpipe/data_types/types.py [hunk 5]
```diff
@@ -469,6 +438,136 @@ def to_ccl(self, **kwargs):
...
+        import tables_io
+        if self._metadata is not None:
+            return self._metadata
+        meta = tables_io.io.readHdf5GroupToDict(self.file["meta"])
+        self._metadata = meta
+        return meta
+
+class QPPDFFile(QPBaseFile):
+    def iterate(self, chunk_rows, rank=0, size=1):
+        import qp
```
Comment (chihway) [inline txpipe/data_types/types.py:459]: dumb questions, why is this better to import in the functions?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -272,7 +272,6 @@ jobs:
         ceci --dry-run examples/2.2i/pipeline.yml
         ceci --dry-run examples/buzzard/pipeline.yml
         ceci --dry-run examples/cosmodc2/pipeline.yml
-        ceci --dry-run examples/dr1b/pipeline.yml
```
Comment (chihway) [inline .github/workflows/ci.yml:275]: why did this get dropped?

File: examples/skysim/config.yml [hunk 7]
```diff
@@ -24,6 +24,27 @@ TXIngestRedmagic:
 TXSourceTrueNumberDensity:
     nz: 301
     zmax: 3.0
+    name: TXSourceTrueNumberDensity
```
Comment (chihway) [inline examples/skysim/config.yml:27]: this is throughout, and Joe said it's a bug, might be a different PR tho

File: txpipe/data_types/types.py [hunk 8]
```diff
@@ -469,6 +438,136 @@ def to_ccl(self, **kwargs):
...
+        if pdf_name == "interp":
+            z = meta["xvals"][:]
+        elif pdf_name == "hist":
+            z = meta["bins"][:]
+        else:
+            raise ValueError(f"TXPipe cannot understand pdf_name in QP file: {pdf_name}")
+        return z.squeeze()
+    
+    
+class QPMultiFile(HDFFile):
```
Comment (chihway) [inline txpipe/data_types/types.py:544]: some comment explaining what this class is

**Comment by chihway** [comment]
Ok @joezuntz I've done a pass, feel free to take or leave whatever I commented on!

File: txpipe/photoz_stack.py (hunk 3 above)
Comment (joezuntz) [reply txpipe/photoz_stack.py:128]: I'm adding a config option to let people choose a different column here.  Using QP to generate draws would have to be from a different class though as the input would need to be a a QP file instead of a plain catalog. We can do that if needed though.

File: examples/lensfit/pipeline.yml (hunk 4 above)
Comment (joezuntz) [reply examples/lensfit/pipeline.yml:103]: Adding a comment!

File: txpipe/data_types/types.py (hunk 5 above)
Comment (joezuntz) [reply txpipe/data_types/types.py:459]: You mean inside the function instead of at top level? That's to avoid loading QP until it's actually needed. It takes a while to load as it has so many dependencies. Also I avoid importing stuff until runtime throughout TXPipe for the same reason, and so that the launcher env can be simpler.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (joezuntz) [reply .github/workflows/ci.yml:275]: It was just so old at this point, and superseded by 2.2i.

File: txpipe/rail/summarize.py (hunk 2 above)
Comment (joezuntz) [reply txpipe/rail/summarize.py]: Done.

File: examples/skysim/config.yml (hunk 7 above)
Comment (joezuntz) [reply examples/skysim/config.yml:27]: Yep - needs a fix in ceci!

Comment (joezuntz) [reply examples/skysim/config.yml:27]: Made an issue here: https://github.com/LSSTDESC/ceci/issues/100

File: txpipe/data_types/types.py (hunk 8 above)
Comment (joezuntz) [reply txpipe/data_types/types.py:554]: done!

**Comment by joezuntz** [comment]
Thanks for reviewing @chihway ! This is merged now.
